File: 1964.find-the-longest-valid-obstacle-course-at-each-position.cs
/*
 * @lc app=leetcode id=1964 lang=csharp
 *
 * [1964] Find the Longest Valid Obstacle Course at Each Position
 */

// @lc code=start
using System.Collections.Generic;
public class Solution 
{
    public static int bisectRight(List<int> nums, int target)
    {
        int left = 0;
        int right = nums.Count;
        while (left < right)
        {
            int mid = left + (right - left) / 2;

            if (nums[mid] <= target)
            {
                left = mid + 1;
            }
            else
            {
                right = mid;
            }
        }
        return left;
    }

    public int[] LongestObstacleCourseAtEachPosition(int[] obstacles) 
    {
        List<int> stack = new List<int>();
        List<int> result = new List<int>();
        foreach (int o in obstacles)
        {
            int i = bisectRight(stack, o);
            result.Add(i + 1);
            if (i == stack.Count)
            {
                stack.Add(o);
            }
            else
            {
                stack[i] = o;
            }
        }
        return result.ToArray();
    }
}

// @lc code=end

File: 1964.find-the-longest-valid-obstacle-course-at-each-position.py
#
# @lc app=leetcode id=1964 lang=python3
#
# [1964] Find the Longest Valid Obstacle Course at Each Position
#

# @lc code=start
import bisect
from typing import List


class Solution:
    '''
    Use LIS monotonic stack method
    '''
    def longestObstacleCourseAtEachPosition(self, obstacles: List[int]) -> List[int]:
        stack = []
        result = []
        for o in obstacles:
            i = bisect.bisect_right(stack, o)
            result.append(i + 1)
            if i == len(stack):
                stack.append(o)
            else:
                stack[i] = o
        return result

# @lc code=end

